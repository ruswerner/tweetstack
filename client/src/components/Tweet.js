import React, {PureComponent} from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import {ListItem} from 'material-ui/List';
import Avatar from 'material-ui/Avatar';
import TweetRecord from '../records/tweet';

export default class Tweet extends PureComponent {

  static propTypes = {
    tweet: PropTypes.instanceOf(TweetRecord).isRequired
  };

  render() {

    const {id, text, user, created_at} = this.props.tweet;

    const ago = moment(new Date(created_at)).fromNow();

    return (
      <ListItem
        onClick={() =>
          window.open(`https://twitter.com/statuses/${id}`)
        }
        leftAvatar={<Avatar src={user.profile_image_url}/>}
        rightIcon={<span>{ago}</span>}
        primaryText={user.name}
        secondaryText={
          /* Tweet text can contain encoded html entities, so this makes it readable.
             ...and we trust that twitter has sanitised the content */
          <div style={{height: 'auto', WebkitLineClamp: 'unset'}} dangerouslySetInnerHTML={{__html: text}}/>
        }
        secondaryTextLines={2}
      >
      </ListItem>
    );
  }

}
