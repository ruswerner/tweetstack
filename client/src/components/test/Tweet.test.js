import React from 'react';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import Tweet from '../Tweet';
import TweetRecord from '../../records/tweet';
import TweetUserRecord from '../../records/tweet-user';

describe('Tweet', function() {

  it('should render', function() {
    const tree = renderer.create(
      <MuiThemeProvider>
        <Tweet
          tweet={new TweetRecord({
            id:'1',
            text:'random text',
            created_at: 'Mon Sep 24 03:35:21 +0000 2012',
            user: new TweetUserRecord({
              name: 'name1',
              screen_name: 'sname1',
              profile_image_url: 'http://_.com/img1.jpg'
            })
          })}
        />
      </MuiThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('when clicked', function() {
    it('should open permalink', function() {
      const spy = jest.spyOn(window, 'open');

      const wrapper = mount(
        <MuiThemeProvider>
          <Tweet
            tweet={new TweetRecord({
              id:'1',
              text:'random text',
              created_at: 'Mon Sep 24 03:35:21 +0000 2012',
              user: new TweetUserRecord({
                name: 'name1',
                screen_name: 'sname1',
                profile_image_url: 'http://_.com/img1.jpg'
              })
            })}
          />
        </MuiThemeProvider>
      );
      wrapper.find('span').first().simulate('click');
      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls[0][0]).toEqual('https://twitter.com/statuses/1');

      spy.mockRestore();
    });
  });


});
