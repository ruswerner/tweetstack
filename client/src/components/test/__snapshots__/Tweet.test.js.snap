// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet should render 1`] = `
<div>
  <span
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "background": "none",
        "backgroundColor": null,
        "border": 10,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": 16,
        "fontWeight": "inherit",
        "lineHeight": "16px",
        "margin": 0,
        "muiPrepared": true,
        "outline": "none",
        "padding": 0,
        "position": "relative",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "verticalAlign": null,
      }
    }
    tabIndex={0}
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <div
        style={
          Object {
            "marginLeft": 0,
            "muiPrepared": true,
            "paddingBottom": 16,
            "paddingLeft": 72,
            "paddingRight": 56,
            "paddingTop": 16,
            "position": "relative",
          }
        }
      >
        <span
          color="#757575"
          style={
            Object {
              "display": "block",
              "height": 24,
              "margin": 12,
              "position": "absolute",
              "right": 4,
              "top": 0,
              "width": 24,
            }
          }
        >
          5 years ago
        </span>
        <img
          className={undefined}
          size={40}
          src="http://_.com/img1.jpg"
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgb(188, 188, 188)",
              "borderRadius": "50%",
              "color": "#ffffff",
              "display": "inline-flex",
              "fontSize": 20,
              "height": 40,
              "justifyContent": "center",
              "left": 16,
              "muiPrepared": true,
              "position": "absolute",
              "top": 16,
              "userSelect": "none",
              "width": 40,
            }
          }
        />
        <div
          style={
            Object {
              "muiPrepared": true,
            }
          }
        >
          name1
        </div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "random text",
            }
          }
          style={
            Object {
              "WebkitBoxOrient": "vertical",
              "WebkitLineClamp": "unset",
              "color": "rgba(0, 0, 0, 0.54)",
              "display": "-webkit-box",
              "fontSize": 14,
              "height": "auto",
              "lineHeight": "18px",
              "margin": 0,
              "marginTop": 4,
              "muiPrepared": true,
              "overflow": "hidden",
              "textOverflow": "ellipsis",
              "whiteSpace": null,
            }
          }
        />
      </div>
    </div>
  </span>
</div>
`;
